File: Data_structures/BST/bst.js
class TreeNode {
    constructor(val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}


class BST {
   constructor() {
       this.root = null;
   }

   insert(val, root = this.root) {
       let newNode = new TreeNode(val);
       if (!this.root) {
           this.root = newNode;
           return;
       }
       if (val < root.val) {
           if (!root.left) {
               root.left = newNode;
           } else {
               return this.insert(val, root.left);
           }
       } else {
           if (!root.right) {
               root.right = newNode;
           } else {
               return this.insert(val, root.right);
           }
       }
   }

   searchRecur(val, root = this.root) {
       if (!root) return false;
       if (val < root.val) {
           return this.searchRecur(val, root.left);
       } else if (val > root.val) {
           return this.searchRecur(val, root.right);
       } else {
           return true;
       }
   }

   searchIter(val, root = this.root) {
       let currentNode = this.root;
       while (currentNode) {
           if (val < currentNode.val) {
               currentNode = currentNode.left;
           } else if (val > currentNode.val) {
               currentNode = currentNode.right;
           } else {
               return true;
           }
       }
       return false;

    //    let queue = [root];
    //    if (!this.root) return false;
    //    while (queue.length) {
    //        let node = queue.shift();
    //        if (val < node.val) {
    //            if (node.left) queue.push(node.left);
    //        } else if (val > node.val) {
    //            if (node.right) queue.push(node.right);
    //        } else {
    //            return true;
    //        }
    //    }
    //    return false;
   }
}

module.exports = {
    TreeNode,
    BST
};

File: python-practice/BST.py
# from .tree_node import TreeNode
# class BST(object):

#     def __init__(self):
#         self.root = None

#     def insert(self, value, root = self.root):
#         if self.root == None:
#             self.root = TreeNode(value)
#         if value < root:
#             if root.left == None:
#                 root.left = TreeNode(value)
#             else:
#                 self.insert(value, root.left)
#         else:
#             if root.right == None:
#                 root.right = TreeNode(value)
#             else:
#                 self.insert(value, root.right)

class TreeNode(object):

    def __init__(self,value):
        self.value = value
        self.left = None
        self.right = None

# def search(root,value):
#     if root is None or root.value == value:
#         return root
#     if (value < root.value):
#         return search(root.left, value)
#     return search(root.right, value)

# def insert(root, value):
#     if root is None:
#         root = TreeNode(value)
#     else:
#         if value < root.value:
#             if root.left is None:
#                 root.left = TreeNode(value)
#             else:
#                 insert(root.left, value)
#         else:
#             if root.right is None:
#                 root.right = TreeNode(value)
#             else:
#                 insert(root.right, value)

# def in_order(root):
#     if root is None:
#         return 
#     else:
#         in_order(root.left)
#         print(root.value)
#         in_order(root.right)

def search(root,value):
    # if root is None:
    #     return False
    # if root.value == value:
    #     return True
    # if (value < root.value):
    #     return search(root.left, value)
    # return search(root.right, value)

    # stack = [root]
    # while len(stack) > 0:
    #     node = stack.pop()
    #     if value < node.value and node.left is not None:
    #         stack.append(node.left)
    #     elif value >= node.value and node.right is not None:
    #         stack.append(node.right)
    #     else:
    #         return True
    node = root
    while node is not None:
        if value < node.value:
            node = node.left
        elif value > node.value:
            node = node.right
        else:
            return True
    return False


def insert(root, value):
    if root is None:
        root = TreeNode(value)
    else:
        if value < root.value:
            if root.left is None:
                root.left = TreeNode(value)
            else:
                insert(root.left, value)
        else:
            if root.right is None:
                root.right = TreeNode(value)
            else:
                insert(root.right, value)

def in_order(root):
    if root is None:
        return 
    else:
        in_order(root.left)
        print(root.value)
        in_order(root.right)



root = TreeNode(50)
insert(root,30)
insert(root,20)
insert(root,40)
insert(root,70)
insert(root,60)
insert(root,80)

in_order(root)
print(search(root,70))
print(search(root,100))